runtime/src/accounts_index.rs

```diff
@@ -174,13 +174,157 @@ impl<T: 'static + Clone + IsCached> WriteAccountMapEntry<T> {
 }
 
 #[derive(Debug, Default)]
+pub struct RollingBitField {
+    max_width: usize,
+    min: u64,
+    max: u64, // exclusive
+    bits: Vec<u64>,
+    count: usize,
+}
+
+#[derive(Debug, Default)]
+struct RollingBitFieldAddress {
+    array_index: usize,
+    mask: u64,
+}
+
+// functionally similar to a hashset
+// Relies on there being a sliding window of key values. The key values continue to increase.
+// Old key values are removed from the lesser values and do not accumulate.
+impl RollingBitField {
+    pub fn new(max_width: usize) -> Self {
+        assert!(max_width > 0);
+        assert!(max_width.is_power_of_two()); // power of 2 to make dividing a shift
+                                              // calculate the array length required using the max width - 1
+        let address = Self::calc_address(max_width, max_width as u64 - 1);
+        let bits = vec![0u64; address.array_index + 1];
+        Self {
+            max_width,
+            bits,
+            count: 0,
+            min: 0,
+            max: 0,
+        }
+    }
+
+    fn get_address(&self, key: u64) -> RollingBitFieldAddress {
+        Self::calc_address(self.max_width, key)
+    }
+
+    // find the array index and mask in the u64 where this key maps
+    fn calc_address(max_width: usize, key: u64) -> RollingBitFieldAddress {
+        let key = key as usize;
+        let bits_in_u64 = 64;
+        let array_index = (key % max_width) / bits_in_u64;
+        let bit_index = key % bits_in_u64;
+        let mask: u64 = 1u64 << bit_index;
+        RollingBitFieldAddress { array_index, mask }
+    }
+
+    pub fn insert(&mut self, key: &u64) {
```

Comment:
any reason not to just take `u64` directly?

Reply:
to match the api that I am replacing. hashset uses &key

Reply:
I agree it is ugly. This was the low impact change to measure, try it out, etc.

Reply:
oops - insert is not a ref